dbms/src/Common/getNumberOfCPUCores.cpp

```diff
@@ -62,8 +64,11 @@ void computeAndSetNumberOfPhysicalCPUCores(
     // - `(hardware_logical_cpu_cores / number_of_hardware_physical_cores)` means how many logical cpu core a physical cpu core has.
     // - `number_of_logical_cpu_cores_ / (hardware_logical_cpu_cores / number_of_hardware_physical_cores)` means how many physical cpu cores the tiflash process could use. (Actually, it's needless to get physical cpu cores in virtual environment, but we must ensure the behavior `1` is not broken)
     auto hardware_logical_cpu_cores = std::thread::hardware_concurrency();
-    UInt16 physical_cpu_cores
-        = number_of_logical_cpu_cores_ / (hardware_logical_cpu_cores / number_of_hardware_physical_cores);
+
+    UInt16 cpu_num_per_physical_core = hardware_logical_cpu_cores / number_of_hardware_physical_cores;
+    cpu_num_per_physical_core = std::max<UInt16>(1, cpu_num_per_physical_core);
+
+    UInt16 physical_cpu_cores = number_of_logical_cpu_cores_ / cpu_num_per_physical_core;
     CPUCores::number_of_physical_cpu_cores = physical_cpu_cores > 0 ? physical_cpu_cores : 1;
```

Comment:
```suggestion
    UInt16 thread_num_per_physical_core = hardware_logical_cpu_cores / number_of_hardware_physical_cores;
    thread_num_per_physical_core = std::max<UInt16>(2, cpu_num_per_physical_core); // ? maybe 2 is better?

    UInt16 physical_cpu_cores = number_of_logical_cpu_cores_ / thread_num_per_physical_core;
    CPUCores::number_of_physical_cpu_cores = std::max<UInt16>(1, physical_cpu_cores);
```

Reply:
> thread_num_per_physical_core

Done. I think 1 is better as there may still be some cpu without hyper-threading.

Reply:
[Hyper-threading](https://en.wikipedia.org/wiki/Hyper-threading) was introduced in 2002.